Fixes zoom not under mouse

HarelM opened the pull request:
## Launch Checklist

Fixes #1024 - Zoom center not under cursor when terrain is on 
While there are a lot of change in this PR, all of them are types I added because I couldn't understand what I'm reading.
The fix is basically the following - where the code related to terrain drag only executes when the map is being dragged and not while scroll zooming.
```
if (!terrain || !combinedEventsInProgress.drag) {
            tr.setLocationAtPoint(loc, around);
```




<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented on src/ui/handler_manager.ts:
```diff
@@ -437,7 +455,7 @@ class HandlerManager {
         if (zoomDelta) tr.zoom += zoomDelta;
 
-        if (!terrain) {
+        if (!terrain || !combinedEventsInProgress.drag) {
```
This is basically the fix, @prozessor13 can you take a look at this?

wipfli commented:
Thanks for working on this, @HarelM. The type improvements can probably be merged as is, but the zoom behavior is probably not idea at the moment. While the zoom center is now under the mouse cursor, the camera altitude now follows terrain. 

When zooming in, it feels like you are in an aeroplane which jumps up and down.

I updated the demo at https://wipfli.github.io/maplibre-terrain3d/swisstopo.html#11.94/46.49885/8.03612/65.6/69 to use your latest code. Feel free to play around with it...

prozessor13 commented:
Hi Harel,

you can try to work with the `transform.freezeElevation` and `transform.recalculateZoom` logic. May this helps with the wobbling effect.

HarelM commented:
Thanks! I'll loo into it.

florianf commented:
Hi @HarelM try the attached patch, this should do the trick.
[zoom.txt](https://github.com/maplibre/maplibre-gl-js/files/10294395/zoom.txt)

wipfli commented:
Thanks @florianf. Here is the content of the file:
```diff
diff --git a/src/ui/handler_manager.ts b/src/ui/handler_manager.ts
index 63296c43cc2f221225406d466932b8e9b52c7612..9d91c72740de24862b1e6106bf078776dd8ac8ac 100644
--- a/src/ui/handler_manager.ts
+++ b/src/ui/handler_manager.ts
@@ -221,6 +221,13 @@ class HandlerManager {
         const scrollZoom = map.scrollZoom = new ScrollZoomHandler(map, this);
         this._add('scrollZoom', scrollZoom, ['mousePan']);
 
+        this._map.on('zoomend', () => {
+            if (this._map.terrain) {
+                this._map.transform.freezeElevation = false;
+                this._map.transform.recalculateZoom(map.terrain);
+            }
+        })
+
         const keyboard = map.keyboard = new KeyboardHandler();
         this._add('keyboard', keyboard);
 
@@ -465,6 +472,10 @@ class HandlerManager {
             } else if (combinedEventsInProgress.drag && this._drag) {
                 tr.center = tr.pointLocation(tr.centerPoint.sub(panDelta));
             }
+            else if (combinedEventsInProgress.zoom) {
+                tr.freezeElevation = true;
+                tr.setLocationAtPoint(loc, around);
+            }
         }
 
         this._map._update();
```

HarelM commented:
Thanks!! Handler manager should not know about zoom related stuff, but thanks for the direction, I'll see how to properly place this code in the right place.

HarelM commented on src/ui/handler_manager.ts:
```diff
@@ -413,12 +430,14 @@ class HandlerManager {
         const terrain = map.terrain;
 
-        if (!hasChange(combinedResult) && !(terrain && this._drag)) {
+        if (!hasChange(combinedResult) && !(terrain && this._terrainMovement)) {
```
@prozessor13 do you remember why you added this part? looks like it should avoid firing events, but I'm not sure why this is needed...?

prozessor13 replied:
Sorry cant remember in detail, but not to avoid firing events. More that the dragging is finished correctly via `freezeElevation = false` and `recalculateZoom`

HarelM commented on src/ui/handler_manager.ts:
```diff
@@ -445,25 +464,22 @@ class HandlerManager {
-                    point: around,
-                    handlerName: combinedEventsInProgress.drag.handlerName
-                };
+            if (!this._terrainMovement && 
```
This is the interesting part now. I've combined the logic of drag and zoom to a single code. I'm not sure this is the right place to do it, but at least it looks the same now.

prozessor13 replied:
The Problem with your code is that `setLocationAtPoint` is not working in terrain. May the method calculates the coordinates ok-ish but, depending on terrain, for shure not correct. I cant remember in detail, but i had a lot of troubles dealing with this method. So i came to the conclusion to let all map drag calculation starting from map center.

HarelM replied:
I left it the safe for drag and only changed zoom. While it might not be accurate, I think the UX is better.
If you have an example I can use I'll be happy to test it.

prozessor13 replied:
today afternoon, when i looked onto my code in handler_manager.ts i realized, that mostly all variables in `this._drag` are no longer used. So i think, mostly all of this logic, is dead code from an earlier implementation :/, and your `terrainMovement` is, apart from using `setLocationAtPoint`, really an simplification/improvement.
May it is sufficent to forward the `terrain` object to `setLocationAtPoint` and the containing `pointCoordinate` methods? ... but i am afraid not.

HarelM commented:
I've rewritten the fix, it's now a bit more complicated, but more "general".
Feel free to review and share your input.
@prozessor13 if you have a few minutes to look at the places I added comments for it would be great.

birkskyum reviewed (approved):
This feels great zooming around.

HarelM commented:
@birkskyum, @wipfli how do you feel about merging this?
I generally think it's better than current UX.
I don't mind revisiting this in the future if need be assuming people will report a problem.
So I think this should be merged.

birkskyum commented:
It's much better than what we had.

wipfli commented:
Happy to try this out at the next opportunity. Do you know if your fix affects performance of non-terrain-3d maps, @HarelM? I guess by performance I mean frame rate when zooming and panning...

HarelM commented:
The code that I changed is not executed in 2D, so there's no performance hit in that case.
